packages/@lwc/ssr-compiler/src/compile-template/transformers/component/slotted-content.ts

```diff
@@ -36,51 +39,58 @@ import type {
 } from '@lwc/template-compiler';
 import type { TransformerContext } from '../../types';
 
-const bGenerateSlottedContent = esTemplateWithYield`
-        const shadowSlottedContent = ${/* hasShadowSlottedContent */ is.literal}
-            ? async function* __lwcGenerateSlottedContent(contextfulParent) {
-                // The 'contextfulParent' variable is shadowed here so that a contextful relationship
-                // is established between components rendered in slotted content & the "parent"
-                // component that contains the <slot>.
-
-                ${/* shadow slot content */ is.statement}
-            } 
-            // Avoid creating the object unnecessarily
-            : null;
-
-        const lightSlottedContentMap = ${/* hasLightSlottedContent */ is.literal} 
-            ? Object.create(null)
-            // Avoid creating the object unnecessarily
-            : null;
-        
-        // The containing slot treats scoped slotted content differently.
-        const scopedSlottedContentMap = ${/* hasScopedSlottedContent */ is.literal} 
-            ? Object.create(null)
-            // Avoid creating the object unnecessarily
-            : null;
-
-        function addSlottedContent(name, fn, contentMap) {
-            let contentList = contentMap[name];
-            if (contentList) {
-                contentList.push(fn);
-            } else {
-                contentMap[name] = [fn];
-            }
-        }
+// This function will be defined once and hoisted to the top of the template function. It'll be
+// referenced deeper in the call stack where the function is called or passed as a parameter.
+// It is a higher-order function that curries local variables that may be referenced by the
+// shadow slot content.
+const bGenerateShadowSlottedContent = esTemplateWithYield`
+    const ${/* function name */ is.identifier} = (${/* local vars */ is.identifier}) => async function* ${/* function name */ 0}(contextfulParent) {
+        // The 'contextfulParent' variable is shadowed here so that a contextful relationship
+        // is established between components rendered in slotted content & the "parent"
+        // component that contains the <slot>.
+        ${/* shadow slot content */ is.statement}
+    };
+`<EsVariableDeclaration>;
+// By passing in the set of local variables (which correspond 1:1 to the variables expected by
+// the referenced function), `shadowSlottedContent` will be curried function that can generate
+// shadow-slotted content.
+const bGenerateShadowSlottedContentRef = esTemplateWithYield`
+    const shadowSlottedContent = ${/* reference to hoisted fn */ is.identifier}(${/* local vars */ is.identifier});
+`<EsVariableDeclaration>;
+const bNullishGenerateShadowSlottedContent = esTemplateWithYield`
+    const shadowSlottedContent = null;
+`<EsVariableDeclaration>;
```

Comment:
(non-blocking) I don't super love having two separate tiny templates that are almost the same thing. If we were to make the template holes in the first one nullable, we could make the generated output `const shadowSlottedContent = (null)?.(null)`. Which is slightly longer but maybe a smart bundler can optimize it? (Also, it sneakily evaluates to `undefined`, not `null`!)

Reply:
I like that better too. I'll make the change.

Reply:
Thought about this more; not loving the direction after all. It does allow us to have a single builder. But the tradeoff is weird generated code. And that doesn't feel like a great trade-off to me. I think what we need is to enhance `esTemplate` such that you can have nested templating with some logic to manage cases like this. I want to give it some more thought and, because you marked this comment as non-blocking, I'll defer that to later.